trace/spans/factory.go

```diff
@@ -0,0 +1,341 @@
+
+	// Sets the span kind to "CONSUMER".  Does nothing except log a failure
+	// with a stack trace if the factory is empty.
+	//
+	SetIsSubscriber()
+
+	// SetDisplayName() sets the display name on the contained span.  Does
+	// nothing except log a failure with a stack trace if the factory is
+	// empty.
+	//
+	SetDisplayName(desc string)
+
+	// AddAttribute() adds an attribute key/value pair to the contained span.
+	// Does nothing except log a failure with a stack trace if the factory is
+	// empty (even returning a 'nil' error).
+	//
+	// 'val' can be a 'string', an 'int' or 'int64', or a 'bool'.  If 'key'
+	// is empty or 'val' is not one of the listed types, then an error is
+	// returned and the attribute is not added.
+	//
+	AddAttribute(key string, val interface{}) error
+
+	// SetStatusCode() sets the status code on the contained span.
+	// 'code' is expected to be a value from
+	// google.golang.org/genproto/googleapis/rpc/code but this is not
+	// verified.  Does nothing except log a failure with a stack trace
+	// if the factory is empty.
+	//
+	SetStatusCode(code int64)
+
+	// SetStatusMessage() sets the status message string on the contained
+	// span.  Does nothing except log a failure with a stack trace if the
+	// factory is empty.
+	//
+	SetStatusMessage(msg string)
+
+	// Finish() notifies the factory that the contained span is finished.
+	// The factory will be empty afterward.  The factory will arrange for the
+	// span to be registered.
+	//
+	// The returned value is the duration of the span's life.  If the factory
+	// was already empty or the contained span was from Import(), then a
+	// failure with a stack trace is logged and a 0 duration is returned.
+	//
+	Finish() time.Duration
+}
+
+func ContextStoreSpan(ctx context.Context, span Factory) context.Context {
+	return context.WithValue(ctx, _contextSpan, span)
+}
+
+func ContextGetSpan(ctx context.Context) Factory {
+	span := ctx.Value(_contextSpan)
+	if nil == span {
+		return nil
+	}
+	return span.(Factory)
+}
+
+// NonHexIndex() returns the offset to the first character in the string that
+// is not a hexadecimal digit (0..9, a..f, A..F) or -1 if none.
+//
+func NonHexIndex(s string) int {
+	for i := 0; i < len(s); i++ {
+		c := s[i]
+		if 0 == HexChars[c>>5]&(1<<uint(c&31)) {
+			return i
+		}
+	}
+	return -1
+}
+
+// IsValidTraceID() returns 'true' only if 's' is a non-zero hexadecimal
+// value of 32 digits.
+//
+func IsValidTraceID(s string) bool {
+	if 32 != len(s) || -1 != NonHexIndex(s) || s == "00000000000000000000000000000000" {
+		return false
+	}
+	return true
+}
+
+func HexSpanID(spanID uint64) string {
+	return fmt.Sprintf("%016x", spanID)
+}
+
+// FinishSpan() calls Finish() on the passed-in 'span', unless it is 'nil'.
+// It is meant to be used with 'defer' when a 'span' might be 'nil':
+//
+//      defer spans.FinishSpan(span)
+//
+func FinishSpan(span Factory) {
+	if nil != span {
+		span.Finish()
+	}
+}
+
+// NewRoSpan() returns an empty Factory.
```

Comment:
What does Ro in RoSpan stand for?

Reply:
Maybe "ROSpan" is better spelling as well?

Reply:
Yeah ROSpan sounds good to me! Or `ReadOnlySpan`? I don't know if that is too verbose for you?

Reply:
Pushed this rename plus explained name.